### PR TITLE
fix: report correct errors for json schema validation

### PR DESCRIPTION
**Pull request by CaptainStandby**
- Implemented the translation of `jsonschema.ValidationError` to errors codes documented [here](https://www.ory.sh/docs/kratos/concepts/ui-user-interface#machine-readable-format)
- Added missing error codes for relevant schema errors
  | Validation         | Name                            | ID      |
  | ------------------ | ------------------------------- | ------- |
  | `maxLength`        | ErrorValidationMaxLength        | 4000017 |
  | `minimum`          | ErrorValidationMinimum.         | 4000018 |
  | `exclusiveMinimum` | ErrorValidationExclusiveMinimum | 4000019 |
  | `maximum`          | ErrorValidationMaximum          | 4000020 |
  | `exclusiveMaximum` | ErrorValidationExclusiveMaximum | 4000021 |
  | `multipleOf`       | ErrorValidationMultipleOf       | 4000022 |
  | `maxItems`         | ErrorValidationMaxItems         | 4000023 |
  | `minItems`         | ErrorValidationMinItems         | 4000024 |
  | `uniqueItems`      | ErrorValidationUniqueItems      | 4000025 |
  | `type`             | ErrorValidationWrongType        | 4000026 |
- Updated e2e tests to check these IDs explicitly

## Related issue(s)

#2974 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
fyi :) in the open source repositories we have a different changelog strategy - here you can use regular commit messages :)

**Comment by CaptainStandby** [comment]
> fyi :) in the open source repositories we have a different changelog strategy - here you can use regular commit messages :)

good to know :-D

File: cmd/clidoc/main.go [hunk 1]
```diff
@@ -80,8 +80,18 @@ func init() {
...
+		"NewErrorValidationMinimum":                               text.NewErrorValidationMinimum("must be >= 5 but found 3"),
+		"NewErrorValidationExclusiveMinimum":                      text.NewErrorValidationExclusiveMinimum("must be > 5 but found 5"),
+		"NewErrorValidationMaximum":                               text.NewErrorValidationMaximum("must be <= 5 but found 6"),
+		"NewErrorValidationExclusiveMaximum":                      text.NewErrorValidationExclusiveMaximum("must be < 5 but found 5"),
+		"NewErrorValidationMultipleOf":                            text.NewErrorValidationMultipleOf("7 not multipleOf 3"),
+		"NewErrorValidationMaxItems":                              text.NewErrorValidationMaxItems("maximum 3 items allowed, but found 4 items"),
+		"NewErrorValidationMinItems":                              text.NewErrorValidationMinItems("minimum 3 items allowed, but found 2 items"),
+		"NewErrorValidationUniqueItems":                           text.NewErrorValidationUniqueItems("items at index 0 and 2 are equal"),
+		"NewErrorValidationWrongType":                             text.NewErrorValidationWrongType("expected number, but got string"),
```
Comment (jonas-jonas) [inline cmd/clidoc/main.go:94]: I guess these need variables, right?

**Review by jonas-jonas — COMMENTED** [review]
Nice work, just one question about the messages & variables.

File: cmd/clidoc/main.go (hunk 1 above)
Comment (jonas-jonas) [reply cmd/clidoc/main.go:94]: ah lol, I see. LGTM then.

**Review by jonas-jonas — DISMISSED** [review]
LGTM.

**Review by aeneasr — DISMISSED** [review]
Epic! Are there any backwards compatibility issues that could happen with this and if so how do we deal with them?

**Comment by aeneasr** [comment]
I see these scenarios:

1. Tests rely on this behavior - this is ok if we break it
2. Custom ui implementation uses translation based on the IDs - it will mean that they have untranslated strings which is ok
3. Custom ui implementation uses the IDs to perform logic - this would break without warning

Did I miss something? What should we do to get (3) over the line smoothly in your opinion?

**Comment by CaptainStandby** [comment]
I don't have a good solution for this.
Maybe we could just issue two errors for the same validation, one with the old value and one with the new. I'm not entirely sure if this is possible and it doesn't sound like the best idea.
WDYT?

**Comment by aeneasr** [comment]
I'm fine with breaking it. I don't think anyone really uses it this way. We can probably check in the community and see if it's a problem before deploying it to production?

**Review by aeneasr — APPROVED** [review]

**Comment by lukaszxion** [comment]
I'm using it. Now there is no context information and the messages look like this:
`Länge muss >= {{expected_length}} sein, hat aber {{actual_length}}.`

The use of `context` was proposed by Ory:
https://github.com/ory/kratos/pull/692#issuecomment-688176348

How can I now get the params for the translations?

**Pull request merged**